src/runtime/objmodel.cpp

```diff
@@ -3293,8 +3301,15 @@ void rearrangeArguments(ParamReceiveSpec paramspec, const ParamNames* param_name
         getArg(kwargs_idx, oarg1, oarg2, oarg3, oargs) = okwargs;
     }
 
-    if ((!param_names || !param_names->takes_param_names) && argspec.num_keywords && !paramspec.takes_kwargs) {
-        raiseExcHelper(TypeError, "%s() doesn't take keyword arguments", func_name);
+    Box* kwargs = nullptr;
+    if (argspec.has_kwargs) {
+        kwargs
+            = getArg(argspec.num_args + argspec.num_keywords + (argspec.has_starargs ? 1 : 0), arg1, arg2, arg3, args);
+    }
+    if ((!param_names || !param_names->takes_param_names) && (argspec.num_keywords || argspec.has_kwargs)
+        && (!paramspec.takes_kwargs
+            && (kwargs && PyDict_Check(kwargs) && static_cast<BoxedDict*>(kwargs)->d.size() > 0))) {
```

Comment:
I don't think this check should ignore the existence of `num_keywords` due to properties of the passed kwargs dict; looks like `dir(a=1)` causes a crash.

I'd be a fan of moving this logic to when we actually try to place a keyword; ie instead of "determine before-hand whether we will have any keywords", we could do "when we try to handle a keyword, if there's nowhere to put it throw the appropriate exception".  Maybe a middle ground is to keep the existing check (for non-kwargs parameters), and then just change the existing kwargs assert to an exception?